Added param to interface pass for Operator functions.

genuinelucifer opened the pull request:
Had to put one extra check for Conversion and Explicit conversion operators...
I don't know if this is okay or we should change the the parameterKind for such operators!

ddobrev commented on src/Generator/Passes/ParamTypeToInterfacePass.cs:
```diff
@@ -7,20 +7,22 @@ public class ParamTypeToInterfacePass : TranslationUnitPass
         {
-            if (!function.IsOperator)
-            {
+            if (!function.IsOperator || function.Parameters.Count > 1)
```
Why have you left this check? What does it have to to with the return type?

genuinelucifer replied:
The unary operators must return the containing type while Binary operators can return other types.
Without this check Build was failing in CSharpTemp.cs at operator ++ in Bar.

ddobrev commented on src/Generator/Passes/ParamTypeToInterfacePass.cs:
```diff
@@ -7,20 +7,22 @@ public class ParamTypeToInterfacePass : TranslationUnitPass
-            if (!function.IsOperator)
-            {
+            if (!function.IsOperator || function.Parameters.Count > 1)
                 ChangeToInterfaceType(function.ReturnType);
```
It's not related to the current issue but now that you're changing the pass anyway, please use function.OriginalReturnType instead.

genuinelucifer replied:
Ok. Will do.

ddobrev commented on src/Generator/Passes/ParamTypeToInterfacePass.cs:
```diff
@@ -7,20 +7,22 @@ public class ParamTypeToInterfacePass : TranslationUnitPass
-                }
-            }
+
+            if(function.OperatorKind != CXXOperatorKind.Conversion &&
```
Please align this whole block.

genuinelucifer replied:
Ok. Will do.

ddobrev replied:
I meant not just the conditional but its body too.

genuinelucifer replied:
:+1:   WIll do in the next commit.

ddobrev commented on src/Generator/Passes/ParamTypeToInterfacePass.cs:
```diff
@@ -7,20 +7,22 @@ public class ParamTypeToInterfacePass : TranslationUnitPass
+
+            if(function.OperatorKind != CXXOperatorKind.Conversion &&
+                                    function.OperatorKind != CXXOperatorKind.ExplicitConversion)
+                     foreach (var parameter in function.Parameters)
```
I don't understand, why have you not added the check for OperatorParameter. If you had a test - please add one - it would've failed.

genuinelucifer replied:
Sorry, i copied few changes between branches and that part didn't got copied. I have corrected it now.

genuinelucifer commented:
@ddobrev I have added the classes and fixed the ParamTypeToInterfacePass code but when writing the test how should I check for the types to be an interface?
Also, it would be much more easy to write test in CSharpTemp.tests because from there i can directly access all the classes. Writing the tests in Generator.Tests is proving dificult.
Can I write the test in CSharpTemp.tests?

ddobrev commented:
@genuinelucifer I was actually going to ask you myself to move the test to CSharpTemp. We only use Generator.Tests when we cannot check what we need at run-time - for example line numbers or dependencies. For this issue you'd get incompilable code if there's a problem so using the regular test projects is enough.

genuinelucifer commented:
@ddobrev Thanks. Will move it. I thought Generator Tests were for all passes. Thanks for the info.

genuinelucifer commented:
@ddobrev Please check it again. I finally merged everything into one commit and added a test too.

ddobrev commented on tests/CSharpTemp/CSharpTemp.h:
```diff
@@ -509,3 +509,27 @@ class DLL_API TestOverrideFromSecondaryBase : public Foo, public SecondaryBase
+{
+	int m;
+public:
+	int getM() { return m; }
```
Could you please move all inlined bodies to the cpp file?

ddobrev commented on tests/CSharpTemp/CSharpTemp.Tests.cs:
```diff
@@ -261,4 +261,20 @@ public void TestNotDestroyingForeignObjects()
+    {
+        var baseClass = new TestParamToInterfacePassBaseTwo();
+        baseClass++;
+        Assert.AreEqual(baseClass.M,1);
```
Please put a space after each comma.

ddobrev commented:
Thanks. I've told @tritao to merge it. I would've done so earlier but you only changed the pull without adding a comment so that I was notified.

genuinelucifer commented:
Oh. Sorry. After doing the changes you told, I rebased both commits to
single. I should have commented. Will remember in future. :)
On 21-Jun-2015 10:18 PM, "Dimitar Dobrev" <notifications@github.com> wrote:

> Thanks. I've told @tritao <https://github.com/tritao> to merge it. I
> would've done so earlier but you only changed the pull without adding a
> comment so that I was notified.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mono/CppSharp/pull/495#issuecomment-113927023>.
>

ddobrev commented:
@genuinelucifer @tritao 's (finally :)) restoring the build server and he'd like to complete that before merging this pull. So please move on to https://github.com/mono/CppSharp/issues/402 and he'll merge this when done.

genuinelucifer commented:
Great! I will start the other issue.

tritao commented on tests/CSharpTemp/CSharpTemp.h:
```diff
@@ -509,3 +509,27 @@ class DLL_API TestOverrideFromSecondaryBase : public Foo, public SecondaryBase
+
+class DLL_API TestParamToInterfacePassBaseTwo
+{
+	int m;
```
I think you need to initialize m to 0 since you're checking its value later with an assert.

ddobrev replied:
You don't initialise this when using the empty ctor. For some reason it happens to work but we cannot rely on that. Please assign it to 0 in the empty ctor.

genuinelucifer replied:
Done :)
I assumed zero as default value for integers. But, I remember now that it's only for global variables. :P
